037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify goals as well as all the other options like he already does today on a `KafkaRebalance` custom resource.
+The cluster operator will use that configuration for the `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
+
+It would also contain a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56), to specify if auto-approval is desired or not when auto-rebalancing is enabled.
+The template to be used is referenced by a new `spec.kafka.autoRebalance.template` field and the cluster operator uses that template to create a corresponding `KafkaRebalance` custom resource.
+If the reference is missing, the cluster operator can create an “empty” `KafkaRebalance` custom resource (which means the optimization proposal will be based on Cruise Control default configuration). It will behave as the rejected alternative.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+      template: # reference to a KafkaRebalanceTemplate custom resource
+```
+
+### Flow
+
+Auto-rebalancing on scaling up.
+
+1. The user scales up the cluster by increasing the `spec.kafka.replicas` field.
+2. The cluster operator takes care of it by scaling up the cluster as usual, by adding additional brokers.
+3. If `spec.kafka.autoRebalance.mode` contains the `scaling` value, the cluster operator creates a `KafkaRebalance` custom resource.
+   * It will specify `spec.mode` as `add-brokers` and `spec.brokers` with the list of brokers already added to the cluster.
+4. The cluster operator asks Cruise Control to generate an optimization proposal via the REST API.
+5. When the optimization proposal is ready, the cluster operator checks if the user has specified auto-approval or not:
+   * If auto-approval is not specified, the user has to approve the proposal manually as usual.
+   * If auto-approval and criteria are met or they are not specified at all, the cluster operator approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
+   * If auto-approval and criteria are not met, the `KafkaRebalance` custom resource status will reflect, by using a warning condition, that the proposal was not approved automatically and the reason for that.
+
+It could happen that the cluster was scaling up but rebalancing didn’t happen because the optimization proposal was not approved.
+To fix this kind of issue, the user should change the rebalancing configuration and allow the cluster operator to reconcile again (but not scaling up because it was already done) for getting a new optimization proposal and running a rebalance.
```

Comment:
After scaling up but rebalancing in error, the corresponding `KafkaRebalance` would be in `NotReady` state.
On each (next) reconcile, the operator could go through the `spec.kafka.autoRebalance` entry and check if there is any "pending" corresponding `KafkaRebalance` to refresh (when in `NotReady`, you can make changes and add `refresh` annotation for getting a new proposal). Of course, it would mean user changing the rebalancing template to fix the issue.

More problematic could be if the scaling up goes ok and operator crashes before creating the `KafkaRebalance` from the template. On next reconcile, the operator doesn't know that before it ran a scaling up and now it needs to run a related rebalancing. We would need some `autoRebalance` related section in the `status`?

> but you don't want to keep track of it forever in case the user just doesn't fix it.

At this point should the user just remove the `autoRebalance` related rebalancing entry to avoid the operator to reconcile?

Reply:
> After scaling up but rebalancing in error, the corresponding KafkaRebalance would be in NotReady state.
On each (next) reconcile, the operator could go through the spec.kafka.autoRebalance entry and check if there is any "pending" corresponding KafkaRebalance to refresh (when in NotReady, you can make changes and add refresh annotation for getting a new proposal). Of course, it would mean user changing the rebalancing template to fix the issue.

That sounds like a possibly never ending story. I think you want to avoid it and have a clear end of the process for all possible paths.